test/dns-test.js

```diff
@@ -0,0 +1,298 @@
+  maxTimeout: 3000,
+  dnssec: true,
+  servers: [`127.0.0.1:${network.nsPort}`]
+});
+
+describe('DNS Servers', function() {
+  this.timeout(15000);
+
+  before(async () => {
+    await node.open();
+    await nstub.open();
+  });
+
+  after(async () => {
+    await node.close();
+    await nstub.close();
+  });
+
+  describe('Authoritative Resolver', () => {
+    // write each resource json to the tree
+    for (const [hstype, item] of Object.entries(json)) {
+      // DNS RR type
+      const type = types[item.type];
+
+      // assert that the type maps are correct
+      const rtype = Records.types[hstype];
+      // handshake type -> DNS RR type
+      assert.equal(Records.dnsByType[rtype], type);
+      // handshake type -> handshake value (string)
+      assert.equal(Records.typesByVal[rtype], hstype);
+
+      const resource = Resource.fromJSON(item.resource);
+
+      let name;
+      before(async () => {
+        // args: size, height, network
+        name = rules.grindName(5, 1, network);
+        const raw = Buffer.from(name, 'ascii');
+        const nameHash = rules.hashName(raw);
+        // create a namestate to save the data to
+        const ns = new NameState();
+        ns.set(raw, 0);
+        ns.setData(resource.encode());
+
+        // Create a database transaction, as
+        // writing directly to the database is
+        // discouraged. Write to the txn and
+        // then commit it.
+        const txn = node.chain.db.tree.transaction();
+        await txn.insert(nameHash, ns.encode());
+        await txn.commit();
+      });
+
+      it(`should return authenticated ${hstype} record`, async () => {
+        // Certain types of queries require the
+        // name to be formatted with additional data
+        const query = buildQuery(name, resource, type);
+
+        // query the authoritative name server
+        let res = await nstub.lookup(query, type);
+
+        // create the dns response locally
+        let dns = resource.toDNS(query, type);
+
+        // query for the zone signing key
+        const dnskey = await nstub.lookup('.', types.DNSKEY);
+
+        // parse the zone signing key and
+        // key signing key out of the response
+        const {zsk, ksk} = getSigningKeys(dnskey);
+
+        assert(zsk instanceof wire.Record);
+        assert(ksk instanceof wire.Record);
+
+        // validate the signature on the ZSK
+        verifyDNSSEC(dnskey, ksk, types.DNSKEY, '.');
+        // validate the signature over the rrsets
+        verifyDNSSEC(res, zsk, type, query);
+
+        // NOTE: the signatures are not canonical
+        // when the native backend is being used
+        // because it uses OpenSSL which does not
+        // yet use RFC 6979, so nullify the
+        // signatures before comparing them
+        if (bcrypto.native === 2) {
+          dns = nullSig(dns);
+          res = nullSig(res);
+          assert(dns && res);
+        }
+
+        assert.deepEqual(dns.answer, res.answer);
```

Comment:
@tuxcanfly I am no longer checking equality of the dns message header, only the `answer`, `authority` and `additional`. The rest of the message may be different based on the configuration of the dns resolver/server.

Reply:
Sounds good! :+1: